src/main/java/com/mraof/minestuck/item/StructureScannerItem.java

```diff
@@ -21,65 +37,112 @@ public class StructureScannerItem extends Item
+					} else
+					{
+						return null;
+					}
+				}));
+		
+		ItemProperties.register(this, new ResourceLocation(Minestuck.MOD_ID, "powered"),
+				((pStack, pLevel, pEntity, pSeed) -> pStack.hasTag() && pStack.getTag().getBoolean("Powered") ? 1 : 0));
 	}
 	
 	@Override
-	public InteractionResultHolder<ItemStack> use(Level levelIn, Player playerIn, InteractionHand handIn)
+	public InteractionResultHolder<ItemStack> use(Level pLevel, Player pPlayer, InteractionHand pUsedHand)
 	{
-		levelIn.playSound(playerIn, playerIn.getX(), playerIn.getY(), playerIn.getZ(), SoundEvents.UI_BUTTON_CLICK.value(), SoundSource.AMBIENT, 0.8F, 1.3F);
+		pLevel.playSound(pPlayer, pPlayer.getX(), pPlayer.getY(), pPlayer.getZ(), SoundEvents.UI_BUTTON_CLICK.get(), SoundSource.AMBIENT, 0.8F, 1.3F);
 		
-		ItemStack item = playerIn.getItemInHand(handIn);
-		boolean foundItem = false;
+		ItemStack stack = pPlayer.getItemInHand(pUsedHand);
 		
-		if(fuelItem != null)
+		if(CheckFuelItem(pPlayer, pLevel))
 		{
-			ItemStack fuelStack = new ItemStack(fuelItem.get());
+			stack.getOrCreateTag().putBoolean("Powered", !stack.getTag().getBoolean("Powered"));
+		}
+		return InteractionResultHolder.success(stack);
+		
+	}
+	
+	public static boolean isCharged(ItemStack stack)
+	{
+		return stack.getDamageValue() < stack.getMaxDamage();
+	}
+	
+	@Override
+	public void inventoryTick(ItemStack pStack, Level pLevel, Entity pEntity, int pSlotId, boolean pIsSelected)
+	{
+		super.inventoryTick(pStack, pLevel, pEntity, pSlotId, pIsSelected);
+		
+		if(pStack.hasTag() && pStack.getTag().getBoolean("Powered") && pLevel instanceof ServerLevel sLevel)
+		{
+			if(!isCharged(pStack))
+			{
+				pStack.setDamageValue(0);
+			}
+			
+			BlockPos pos = sLevel.findNearestMapStructure(structure, pEntity.blockPosition(), 100, false);
 			
-			for(ItemStack invItem : playerIn.getInventory().items)
+			if(pos == null)
 			{
-				if(ItemStack.isSameItem(invItem, fuelStack))
+				pStack.getTag().remove("TargetLocation");
+			} else
+			{
+				pStack.getTag().put("TargetLocation", NbtUtils.writeBlockPos(pos));
+			}
+			
+			if(pEntity.tickCount % 20 == 0)
+			{
+				pStack.hurt(1, pLevel.random, pEntity instanceof ServerPlayer ? (ServerPlayer) pEntity : null);
+				
+				if(!isCharged(pStack))
 				{
-					foundItem = true;
-					if(!levelIn.isClientSide && levelIn.getRandom().nextFloat() >= 0.95F)
-					{
-						invItem.shrink(1);
-						levelIn.playSound(playerIn, playerIn.getX(), playerIn.getY(), playerIn.getZ(), SoundEvents.BLAZE_SHOOT, SoundSource.AMBIENT, 0.4F, 2F);
-					}
-					
-					break;
+					pStack.getTag().putBoolean("Powered", false);
 				}
 			}
-		} else
-		{
-			foundItem = true;
 		}
+	}
+	
+	public boolean CheckFuelItem(Player pPlayer, Level pLevel)
+	{
+		ItemStack fuelStack = new ItemStack(fuelItem.get());
+		
 		
-		if(levelIn instanceof ServerLevel level)
+		if(fuelItem != null)
```

Comment:
If `fuelItem` is allowed to be null, it should be annotated with `@Nullable`. And if it is, you shouldn't use it to create `fuelStack` before checking if `fuelItem` is null. If it isn't allowed to be null, then this check is unneeded.

Reply:
Alright, annotated with `@Nullable` for both the class item and the constructor's parameter. Also moved the fuelStack initialization inside of the check.